p2p/src/network/kad/bootstrap/p2p_network_kad_bootstrap_reducer.rs

```diff
@@ -64,11 +72,19 @@ impl P2pNetworkKadBootstrapState {
                     .take(requests_to_create) // and stop when we create enough requests so up to 3 will be executed in parallel
                     .collect::<Vec<_>>();
 
+                if bootstrap_state.requests.is_empty() && peer_id_req_vec.is_empty() {
```

Comment:
I would rather have this happen later after `.into_dispatcher_and_state()` gets called (even if that makes us go through the next block of code and a loop that will be a no-op). It is easier to follow the code if there is a clear split between sate updates, and state machine flow control. With this shortcuiting here we have state update + state machine flow control code interleaved.